statsd/client_config.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2012-2016 Eli Janssen
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file.
+
+package statsd
+
+import (
+	"fmt"
+	"time"
+)
+
+type ClientConfig struct {
+	// addr is a string of the format "hostname:port", and must be something
+	// validly parsable by net.ResolveUDPAddr.
+	Address string
+
+	// prefix is the statsd client prefix. Can be "" if no prefix is desired.
+	Prefix string
+
+	// ResInterval is the interval over which the addr is re-resolved.
+	// Do note that this /does/ add overhead!
+	// If you need higher performance, leave unset (or set to 0),
+	// in which case the address will not be re-resolved.
+	//
+	// Note that if Address is an {ip}:{port} and not a {hostname}:{port}, then
+	// ResInterval will be ignored.
+	ResInterval time.Duration
+
+	// UseBuffered determines whether a buffered sender is used or not.
+	// If a buffered sender is /not/ used, FlushInterval and FlushBytes values are
+	// ignored. Default is false.
+	UseBuffered bool
+
+	// FlushInterval is a time.Duration, and specifies the maximum interval for
+	// packet sending. Note that if you send lots of metrics, you will send more
+	// often. This is just a maximal threshold.
+	// If FlushInterval is 0, defaults to 300ms.
+	FlushInterval time.Duration
+
+	// If flushBytes is 0, defaults to 1432 bytes, which is considered safe
+	// for local traffic. If sending over the public internet, 512 bytes is
+	// the recommended value.
+	FlushBytes int
+}
+
+// NewClientWithConfig returns a new BufferedClient
+//
+// config is a ClientConfig, which holds various configuration values.
+func NewClientWithConfig(config *ClientConfig) (Statter, error) {
```

Comment:
Why is ClientConfig a pointer?  Everyone in go just loves pointers :) . I would just use it as a struct to make it clear the values of `config` are unused after `NewClientWithConfig` returns.

Reply:
Pointers reduce copying, so people probably just default to them via memory. The format

```
x := &SomeStruct{}
```

is pretty "muscle memory" too. 

Further, the general community guidelines are "when in doubt, use a pointer" [reference](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type)

Reply:
Those comments are for Receiver types which are these https://tour.golang.org/methods/4

```
func (c * Config) bob() {}
```

vs 
```
func (c Config) bob() {}
```

Using pointer vs non-pointer there is much more ambiguous.

I also bet the go runtime is pretty smart about memory allocation and the struct.  Also, sometimes passing pointers can confuse the Go compiler and force it to make the Config object on the heap, vs copying the values into the stack (where stack is preferred over heap).

(Pointer for Config isn't a blocker, just preference feedback).